server/services/v1/database/secondary_indexer.go

```diff
@@ -0,0 +1,559 @@
+	}
+	for _, field := range q.coll.QueryableFields {
+		if schema.IsReservedField(field.Name()) {
+			continue
+		}
+
+		if field.DataType == schema.ArrayType {
+			newRows, err := q.indexArray(tableData.RawData, field, field.KeyPath())
+			if err != nil {
+				return nil, err
+			}
+			rows = append(rows, newRows...)
+		} else {
+			row, err := q.indexField(tableData.RawData, field.FieldName, field.DataType, 0, field.KeyPath()...)
+			if err != nil {
+				if isIgnoreableError(err) {
+					continue
+				} else {
+					log.Err(err).Msgf("Failed to index field name: %s", field.FieldName)
+					return nil, err
+				}
+			}
+			rows = append(rows, *row)
+		}
+	}
+	return rows, nil
+}
+
+func (q *SecondaryIndexer) buildTSRows(tableData *internal.TableData) ([]IndexRow, error) {
+	var rows []IndexRow
+
+	timeStamps := []struct {
+		ts    *internal.Timestamp
+		field schema.ReservedField
+	}{
+		{
+			tableData.CreatedAt,
+			schema.CreatedAt,
+		},
+		{
+			tableData.UpdatedAt,
+			schema.UpdatedAt,
+		},
+	}
+
+	for _, ts := range timeStamps {
+		val := []byte{}
+		dt := schema.NullType
+		if ts.ts != nil {
+			val = []byte(ts.ts.ToRFC3339())
+			dt = schema.DateTimeType
+		}
+
+		row, err := newIndexRow(dt, q.collation, schema.ReservedFields[ts.field], val, 0, false)
+		if err != nil {
+			return nil, err
+		}
+		if err == nil {
+			rows = append(rows, *row)
+		} else {
+			if isIgnoreableError(err) {
+				continue
+			} else {
+				return nil, err
+			}
+		}
+	}
+	return rows, nil
+}
+
+func (q *SecondaryIndexer) buildIndexKey(row IndexRow, primaryKey []interface{}) keys.Key {
+	version := getFieldVersion(row.name, q.coll)
+	return newKeyWithPrimaryKey(primaryKey, q.coll.EncodedName, q.coll.Indexes.SecondaryIndex.Name, KVSubspace, row.Name(), version, row.value.AsInterface(), row.pos)
+}
+
+func (q *SecondaryIndexer) createKeysAndIndexInfo(primaryKey []interface{}, rows []IndexRow) ([]keys.Key, map[string]int64, map[string]int64) {
+	indexKeys := make([]keys.Key, 0, len(rows))
+	sizeIncrease := int64(0)
+	stubs := map[string]bool{}
+	rowCounts := map[string]int64{}
+	rowSizes := map[string]int64{}
+	for _, row := range rows {
+		// Only add one stub per repeated nested field to indicate this document contains the nested field
+		if row.stub && stubs[row.Name()] {
+			continue
+		} else {
+			stubs[row.Name()] = true
+		}
+
+		indexKey := q.buildIndexKey(row, primaryKey)
+		indexKeys = append(indexKeys, indexKey)
+
+		if val, ok := rowCounts[row.Name()]; ok {
+			rowCounts[row.Name()] = val + 1
+		} else {
+			rowCounts[row.Name()] = 1
+		}
+
+		sizeIncrease += int64(len(indexKey.SerializeToBytes()))
+		if val, ok := rowSizes[row.name]; ok {
+			rowSizes[row.Name()] = val + sizeIncrease
+		} else {
+			rowSizes[row.Name()] = sizeIncrease
+		}
+	}
+	return indexKeys, rowSizes, rowCounts
+}
+
+func (q *SecondaryIndexer) indexField(doc []byte, fieldName string, dataType schema.FieldType, pos int, keyPath ...string) (*IndexRow, error) {
```

Comment:
nit: you can return `*IndexRow, bool, error` so that if a field is not indexable then you can return false instead of error but you don't need to do it in this diff. This can be combined when we have an attribute on the field to tell us whether the field needs to be indexed.

Reply:
Awesome. Nice idea